### ISSUE TITLE
Adding M1-Max result

### ISSUE THREAD

**Issue by thipokKub** [rasbt/machine-learning-notes#3]
I ran your script and got the following result

My machine is M1-Max 32c 64 GB, approximate Ram usage `25`-`30` GB
```
torch 1.12.0.dev20220518
device mps
Epoch: 001/001 | Batch 0000/1406 | Loss: 2.3857
Epoch: 001/001 | Batch 0100/1406 | Loss: 2.4062
Epoch: 001/001 | Batch 0200/1406 | Loss: 2.1027
Epoch: 001/001 | Batch 0300/1406 | Loss: 2.0253
Epoch: 001/001 | Batch 0400/1406 | Loss: 2.1160
Epoch: 001/001 | Batch 0500/1406 | Loss: 1.9523
Epoch: 001/001 | Batch 0600/1406 | Loss: 1.9365
Epoch: 001/001 | Batch 0700/1406 | Loss: 2.3179
Epoch: 001/001 | Batch 0800/1406 | Loss: 1.9971
Epoch: 001/001 | Batch 0900/1406 | Loss: 1.7516
Epoch: 001/001 | Batch 1000/1406 | Loss: 1.8922
Epoch: 001/001 | Batch 1100/1406 | Loss: 1.8546
Epoch: 001/001 | Batch 1200/1406 | Loss: 1.7630
Epoch: 001/001 | Batch 1300/1406 | Loss: 1.8767
Epoch: 001/001 | Batch 1400/1406 | Loss: 1.5391
Time / epoch without evaluation: 42.28 min
Epoch: 001/001 | Train: 0.00% | Validation: 0.00% | Best Validation (Ep. 001): 0.00%
Time elapsed: 48.54 min
Total Training Time: 48.54 min
Test accuracy 0.00%
Total Time: 49.99 min
```

**Comment by rasbt** [comment]
Nice! Thanks! I am actually currently updating the benchmarks adding inference speeds, and I will add yours too (with acknowledgement)

**Comment by rasbt** [comment]
![Screen Shot 2022-05-20 at 9 35 19 AM](https://user-images.githubusercontent.com/5618407/169550947-ba58d2ba-e3da-4d79-8ba3-25fabe17b37e.png)
![Screen Shot 2022-05-20 at 9 35 22 AM](https://user-images.githubusercontent.com/5618407/169550954-13ce9ace-f5c5-43d1-a5c5-d28f623189af.png)

**Comment by dangrie158** [comment]
Maybe interesting for you: I just maxed out my M1 Max with a batch size of 128 and got the following:
<img width="339" alt="Screenshot 2022-05-21 at 15 47 16" src="https://user-images.githubusercontent.com/2911800/169654849-cb5f971c-d10b-438e-9983-86ec0e65905a.png">

I guess a 128GB MacBook could fit a batch size of 256

**Comment by rasbt** [comment]
Nice, thanks for sharing! I also got slightly faster results with batch size 64 (instead of the original 32) on the M1 Pro. But yeah, would have to increase the batch size for the Nvidia cards and regular M1 then as well to make it a fair comparison.

**Comment by lutchu** [comment]
> Test accuracy 0.00%
Is this expected?

**Comment by rasbt** [comment]
Yeah it's weird but it happened to me to. 

E.g,. on the CPU it's fine: 
- https://github.com/rasbt/machine-learning-notes/blob/main/benchmark/pytorch-m1-gpu/vgg16-cifar10-results/m1pro-cpu.txt

On the GPU, on the exact same machine with the exact same code, I also got this issue:
- https://github.com/rasbt/machine-learning-notes/blob/main/benchmark/pytorch-m1-gpu/vgg16-cifar10-results/m1pro-gpu.txt

**Comment by thipokKub** [comment]
Could you tried the new nightly build? From https://github.com/pytorch/pytorch/issues/77753, there is a new nightly version which seems to fix the problem (the RAM usage did not grow as much as before)

**Comment by thipokKub** [comment]
> > Test accuracy 0.00%
> 
> Is this expected?

In `compute_accuracy` I think the result should be convert to the same device first like the following
```python
correct_pred += (predicted_labels.cpu() == targets.cpu()).sum()
```

This result in non-zero accuracy (at least for me)

**Comment by rasbt** [comment]
The original code works fine on the NVIDIA GPU, but good catch, I will give it a try later!

**Comment by rasbt** [comment]
@thipokKub there was a new nightly release that fixed the RAM leak issue. I was rerunning the experiments and it was actually quite a bit faster (currently putting together the new results). No pressure, but in case you have time and want to run the M1 Max results again, that'd be cool

**Comment by thipokKub** [comment]
Already did @rasbt

**Comment by rasbt** [comment]
Oh, how did I not see that 😅! Thanks!

**Comment by rasbt** [comment]
Alright, just updated the results at https://sebastianraschka.com/blog/2022/pytorch-m1-gpu.html

**Comment by thipokKub** [comment]
pytorch `1.13.0.dev20220610` seems to get significantly faster (-10 minutes)
Some observation are
- I saw huge spike in GPU power consumption when plugging in using 94w charger (change from ~21w to ~28w on average)
- The memory bandwidth seems to increase from ~45 GB/s to ~75 GB/s

```
torch 1.13.0.dev20220610
device mps
Epoch: 001/001 | Batch 0000/1406 | Loss: 2.5373
Epoch: 001/001 | Batch 0100/1406 | Loss: 2.0927
Epoch: 001/001 | Batch 0200/1406 | Loss: 2.1096
Epoch: 001/001 | Batch 0300/1406 | Loss: 2.0650
Epoch: 001/001 | Batch 0400/1406 | Loss: 1.8195
Epoch: 001/001 | Batch 0500/1406 | Loss: 1.9852
Epoch: 001/001 | Batch 0600/1406 | Loss: 2.0264
Epoch: 001/001 | Batch 0700/1406 | Loss: 1.9916
Epoch: 001/001 | Batch 0800/1406 | Loss: 1.9276
Epoch: 001/001 | Batch 0900/1406 | Loss: 1.8869
Epoch: 001/001 | Batch 1000/1406 | Loss: 2.0278
Epoch: 001/001 | Batch 1100/1406 | Loss: 1.9551
Epoch: 001/001 | Batch 1200/1406 | Loss: 1.7823
Epoch: 001/001 | Batch 1300/1406 | Loss: 1.7606
Epoch: 001/001 | Batch 1400/1406 | Loss: 1.9401
Time / epoch without evaluation: 20.43 min
Epoch: 001/001 | Train: 33.36% | Validation: 33.16% | Best Validation (Ep. 001): 33.16%
Time elapsed: 24.89 min
Total Training Time: 24.89 min
Test accuracy 33.11%
Total Time: 25.96 min
```

I am not sure, but maybe with macOS ventura release (with Metal 3), the performance can be further improved?

**Comment by rasbt** [comment]
Wow thanks! I should do an update of this some time! Haha, I may wait until the macOS Ventura release in Fall though.

**Comment by jonasmerlin** [comment]
Are there any updates on the performance of the M1 Max with Ventura and newer versions of PyTorch?

**Comment by callowaysutton** [comment]
Update:
```
torch 2.4.0.dev20240501
device mps
Epoch: 001/001 | Batch 0000/1406 | Loss: 2.4635
Epoch: 001/001 | Batch 0100/1406 | Loss: 2.1030
Epoch: 001/001 | Batch 0200/1406 | Loss: 1.9349
Epoch: 001/001 | Batch 0300/1406 | Loss: 1.9248
Epoch: 001/001 | Batch 0400/1406 | Loss: 2.4663
Epoch: 001/001 | Batch 0500/1406 | Loss: 1.6643
Epoch: 001/001 | Batch 0600/1406 | Loss: 1.8012
Epoch: 001/001 | Batch 0700/1406 | Loss: 1.9272
Epoch: 001/001 | Batch 0800/1406 | Loss: 1.7155
Epoch: 001/001 | Batch 0900/1406 | Loss: 1.9375
Epoch: 001/001 | Batch 1000/1406 | Loss: 1.9579
Epoch: 001/001 | Batch 1100/1406 | Loss: 1.7168
Epoch: 001/001 | Batch 1200/1406 | Loss: 1.6184
Epoch: 001/001 | Batch 1300/1406 | Loss: 1.9197
Epoch: 001/001 | Batch 1400/1406 | Loss: 1.6845
Time / epoch without evaluation: 18.25 min
Epoch: 001/001 | Train: 34.86% | Validation: 36.02% | Best Validation (Ep. 001): 36.02%
Time elapsed: 21.98 min
Total Training Time: 21.98 min
Test accuracy 35.97%
Total Time: 22.90 min

**Comment by rasbt** [comment]
Thanks for sharing! I am surprised to see the decline in prediction accuracy here similar to what the previous code had above. Maybe it's a PyTorch version change that triggered that; maybe need to investigate and try different hparam settings here.
